### PR TITLE
Expand supported parameters of schema and config classes

### PR DESCRIPTION
**Pull request by hanwen-pcluste**
Major
1. Expand supported parameters. The overall progress can be seen from slurm.simple.yaml. The file is expanding as the development progresses. Finally, slurm.full.yaml will be properly handled.

Minors
1. Add `on_bind_field()` to BaseSchema. With the method, we don't have to specify `data_key` everytime we define a field as long as the CamelCase name defined in the config file can be converted to snake_case in the code.
2. Reordered the classes to split the `cluster_schema.py` and `cluster_config.py` into Storage, Networking, Nodes, Monitoring, Others, Root sections.
3. Although we can use `@validate_schema` to validate some of the semantic rules in marshmallow layer, we decided to only do syntactic validation in marshmallow layer and do everything else in the config layer.
4. Two approaches of writing config validators are under discussion. Here we try out the two approaches. For EBS config，separated validators are used and the order of the validation is managed by `priority` in `cluster_config.py`. On the contrary, for FSx config, although there are many validation methods, they are orchestrated by the only one public method to interact with `cluster_config.py`. In the second approach the orchestrator takes care of order and dependencies of the validation methods.
5. Unit tests are far from enough and are to be added in the future.
6. The messages from the validators should be amended.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: cli/src/pcluster/config/cluster_config.py [hunk 1]
```diff
@@ -159,66 +205,79 @@ def __init__(self, queues_config: List[QueueConfig], scheduler: str = "slurm"):
...
+        if iops is None:
+            if volume_type in EBS_VOLUME_TYPE_IOPS_DEFAULT:
+                iops = EBS_VOLUME_TYPE_IOPS_DEFAULT.get(volume_type)
         self.iops = iops
         self.size = size
         self.encrypted = encrypted
         self.kms_key_id = kms_key_id
         self.snapshot_id = snapshot_id
         self.id = id
+        if throughput is None:
```
Comment (enrico-usai) [inline cli/src/pcluster/config/cluster_config.py:243]: minor: `if throughput is None and volume_type == "gp3":`

File: cli/src/pcluster/schemas/cluster_schema.py [hunk 2]
```diff
@@ -151,17 +237,30 @@ def make_config(self, data, **kwargs):
         return SchedulingConfig(**data)
 
 
+class RaidSchema(BaseSchema):
+    """Represent the schema of the parameters specific to Raid. It is the child of EBS schema."""
+
+    raid_type = fields.Str(data_key="Type", validate=validate.OneOf(["0", "1"]))
+    number_of_volumes = fields.Int(data_key="NumberOfVolumes", missing=2, validate=validate.Range(min=2, max=5))
```
Comment (enrico-usai) [inline cli/src/pcluster/schemas/cluster_schema.py:244]: to be removed?

File: cli/src/pcluster/schemas/cluster_schema.py [hunk 3]
```diff
@@ -170,25 +269,88 @@ def make_config(self, data, **kwargs):
...
+
+    storage_capacity = fields.Int(data_key="StorageCapacity")
+    deployment_type = fields.Str(
+        data_key="DeploymentType", validate=validate.OneOf(["SCRATCH_1", "SCRATCH_2", "PERSISTENT_1"])
+    )
+    imported_file_chunk_size = fields.Int(data_key="ImportedFileChunkSize")
+    export_path = fields.Str(data_key="ExportPath")
+    import_path = fields.Str(data_key="ImportPath")
+    weekly_maintenance_start_time = fields.Str(
+        data_key="WeeklyMaintenanceStartTime", validate=validate.Regexp(r"NONE|^[1-7]:([01]\d|2[0-3]):([0-5]\d)$")
```
Comment (enrico-usai) [inline cli/src/pcluster/schemas/cluster_schema.py:309]: let's remove NONE from all the params

File: cli/src/pcluster/schemas/cluster_schema.py [hunk 4]
```diff
@@ -199,65 +361,178 @@ def make_config(self, data, **kwargs):
...
+    def make_config(self, data, **kwargs):
+        """Generate config object."""
+        return ClusterConfig(**data)
+
+    '''
     @pre_load
     def move_image(self, input_data, **kwargs):
         """Move image field into the head node."""
         # If image is not present in the Head node we can use the one from the Cluster
         input_data["HeadNode"]["Image"] = input_data["Image"]
```
Comment (enrico-usai) [inline cli/src/pcluster/schemas/cluster_schema.py:536]: How did you manage this?

File: cli/src/pcluster/schemas/validators.py [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+
+#
+# This module contains all the classes representing the Schema of the configuration file.
+# These classes are created by following marshmallow syntax.
+#
+from pcluster.config.validators import FSX_MESSAGES
+from pcluster.constants import FSX_HDD_THROUGHPUT, FSX_SSD_THROUGHPUT
+
+
+def efs_existence_of_mode_throughput_validator(throughput_mode, provisioned_throughput):
```
Comment (enrico-usai) [inline cli/src/pcluster/schemas/validators.py:20]: This code seems simple enough to reside in the EFSSchema

File: cli/tests/pcluster/validators/test_validators.py [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+            {"VolumeType": "gp3", "Throughput": 760, "Iops": 3000},
+            "Throughput to IOPS ratio of .* is too high",
+        ),
+    ],
+)
+def test_ebs_volume_throughput_validator(section_dict, expected_error):
+    with pytest.raises(ConfigValidationError, match=expected_error):
+        ebs_config = EbsSchema().load(section_dict)
+        ebs_config.raise_on_error = True
+        ebs_config.validate()
```
[the remark this answers is not in the document]
Comment (hanwen-pcluste) [reply cli/tests/pcluster/validators/test_validators.py]: This depends on further decisions about if there should be a central validators handling other validators. 
If we finally decide to split huge validators to small ones, we need to change the test code to be granular.
If we decide to have a central validator for each section, the current code can be kept.

File: cli/src/pcluster/schemas/cluster_schema.py [hunk 7]
```diff
@@ -45,32 +105,58 @@ def evaluate_dynamic_defaults(self, raw_data, **kwargs):
                 raw_data[fieldname] = field.metadata.get("dynamic_default")(raw_data)
         return raw_data
 
+    @pre_dump
+    def remove_implied_values(self, data, **kwargs):
+        """Remove value implied by the code. i.e., only keep parameters that were specified in the yaml file."""
+        _remove_implied_values(data)
```
[the remark this answers is not in the document]
Comment (hanwen-pcluste) [reply cli/src/pcluster/schemas/cluster_schema.py]: Great catch. After a closer look, the recursive is not needed at all. Marshmallow handles the tree structure for us.

File: cli/src/pcluster/schemas/validators.py [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+
+def validate_storage_type_options(
+    errors, fsx_storage_type, fsx_deployment_type, fsx_per_unit_storage_throughput, fsx_drive_cache_type
+):
+    if fsx_storage_type == "HDD":
+        if fsx_deployment_type != "PERSISTENT_1":
+            errors.append("For HDD filesystems, 'deployment_type' must be 'PERSISTENT_1'")
+        if fsx_per_unit_storage_throughput not in FSX_HDD_THROUGHPUT:
+            errors.append(
+                "For HDD filesystems, 'per_unit_storage_throughput' can only have the following values: {0}".format(
```
[the remark this answers is not in the document]
Comment (enrico-usai) [reply cli/src/pcluster/schemas/validators.py:128]: I mean we cannot talk about `per_unit_storage_throughput` because now the parameter is `PerUnitStorageThroughput`. To avoid this misalignements we should show an error message like:

`For HDD filesystems, the per unit storage throughput can only have the following values...` or something like this.

File: cli/src/pcluster/config/cluster_config.py [hunk 9]
```diff
@@ -159,66 +222,78 @@ def __init__(self, queues_config: List[QueueConfig], scheduler: str = "slurm"):
...
         self.kms_key_id = kms_key_id
         self.snapshot_id = snapshot_id
         self.id = id
+        if throughput is None and volume_type == "gp3":
+            throughput = MyInt(125)
+        self.throughput = throughput
+        self.raid = raid
 
+        self._register_validator(
+            EbsVolumeTypeSizeValidator, dependency=2, volume_type=self.volume_type, volume_size=self.size
```
Comment (yuleiwan) [inline cli/src/pcluster/config/cluster_config.py:266]: For the dependency, is there any reason to start from 2? I am thinking we should start from middle of the range so we can extend the validator in future and needn't change previous dependency number.

File: cli/src/pcluster/config/cluster_config.py [hunk 10]
```diff
@@ -277,6 +353,160 @@ def __init__(
...
+        mount_dir: str,
+        ebs_settings: EbsConfig = None,
+        efs_settings: EfsConfig = None,
+        fsx_settings: FsxConfig = None,
+    ):
+        super().__init__()
+        self.mount_dir = mount_dir
+        if ebs_settings:
+            self.ebs_settings = ebs_settings
+        elif efs_settings:
```
Comment (yuleiwan) [inline cli/src/pcluster/config/cluster_config.py:370]: Does the cluster allow multiple file systems? Since I see the test template `slurm.full.yaml` contains 3 file systems, here just enables to assign one to cluster

**Review by yuleiwan — COMMENTED** [review]
Nice work!

File: cli/src/pcluster/config/cluster_config.py (hunk 10 above)
Comment (yuleiwan) [reply cli/src/pcluster/config/cluster_config.py:370]: I see this is defined in cluster schema as many, never mind

File: cli/src/pcluster/config/cluster_config.py [hunk 11]
```diff
@@ -14,48 +14,94 @@
...
-import operator
 from abc import ABC
-from enum import Enum
 from typing import List
 
+from pcluster.constants import EBS_VOLUME_TYPE_IOPS_DEFAULT
 from pcluster.validators.common import ValidationResult, Validator
 from pcluster.validators.ec2 import InstanceTypeValidator
 from pcluster.validators.fsx import FsxS3OptionsValidator
+from pcluster.validators.validators import (
```
[the remark this answers is not in the document]
Comment (hanwen-pcluste) [reply cli/src/pcluster/config/cluster_config.py]: Done

File: cli/src/pcluster/validators/validators.py [hunk 12]
```diff
@@ -0,0 +1,95 @@
+# Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (hanwen-pcluste) [reply cli/src/pcluster/validators/validators.py]: Done

File: cli/src/pcluster/schemas/cluster_schema.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (hanwen-pcluste) [reply cli/src/pcluster/schemas/cluster_schema.py]: Done

File: cli/src/pcluster/schemas/cluster_schema.py [hunk 13]
```diff
@@ -170,25 +269,88 @@ def make_config(self, data, **kwargs):
...
-    type = fields.Str(data_key="StorageType")
-    provisioned_throughput = fields.Str(data_key="ProvisionedThroughput")
-    # TODO add missing fields
+    encrypted = fields.Bool(data_key="Encrypted")
+    kms_key_id = kms_key_id()
+    performance_mode = fields.Str(data_key="PerformanceMode", validate=validate.OneOf(["generalPurpose", "maxIO"]))
+    throughput_mode = fields.Str(data_key="ThroughputMode", validate=validate.OneOf(["provisioned", "bursting"]))
+    provisioned_throughput = fields.Int(data_key="ProvisionedThroughput", validate=validate.Range(min=0, max=1024))
+    file_system_id = fields.Str(
+        data_key="FileSystemId", validate=validate.Regexp(r"^fs-[0-9a-z]{8}$|^fs-[0-9a-z]{17}|NONE$")
```
[the remark this answers is not in the document]
Comment (hanwen-pcluste) [reply cli/src/pcluster/schemas/cluster_schema.py]: Done

File: cli/src/pcluster/config/cluster_config.py (hunk 1 above)
Comment (hanwen-pcluste) [reply cli/src/pcluster/config/cluster_config.py]: Done

File: cli/src/pcluster/schemas/cluster_schema.py (hunk 2 above)
Comment (hanwen-pcluste) [reply cli/src/pcluster/schemas/cluster_schema.py]: Yes, done

File: cli/src/pcluster/schemas/cluster_schema.py (hunk 3 above)
Comment (hanwen-pcluste) [reply cli/src/pcluster/schemas/cluster_schema.py]: Done

File: cli/tests/pcluster/validators/test_validators.py [hunk 14]
```diff
@@ -0,0 +1,154 @@
+# Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (hanwen-pcluste) [reply cli/tests/pcluster/validators/test_validators.py]: Done

File: cli/src/pcluster/schemas/validators.py (hunk 5 above)
Comment (hanwen-pcluste) [reply cli/src/pcluster/schemas/validators.py:18]: Done

File: cli/tests/pcluster/schemas/test_validators.py [hunk 15]
```diff
@@ -0,0 +1,93 @@
+# Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (hanwen-pcluste) [reply cli/tests/pcluster/schemas/test_validators.py:1]: Done

File: cli/src/pcluster/schemas/cluster_schema.py [hunk 16]
```diff
@@ -13,24 +13,84 @@
...
+    "subnet_id": r"^subnet-[0-9a-z]{8}$|^subnet-[0-9a-z]{17}$",
+    "volume_id": r"^vol-[0-9a-z]{8}$|^vol-[0-9a-z]{17}$",
+    "volume_types": ["standard", "io1", "io2", "gp2", "st1", "sc1", "gp3"],
+    "architectures": SUPPORTED_ARCHITECTURES,
+    "fsx_storage_types": ["SSD", "HDD"],
+    "fsx_drive_cache_types": ["READ"],
+}
+
+
+def get_validator(field_name):
```
[the remark this answers is not in the document]
Comment (hanwen-pcluste) [reply cli/src/pcluster/schemas/cluster_schema.py]: Done

File: cli/tests/pcluster/schemas/test_schema_validators.py [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+            {"ThroughputMode": "bursting", "ProvisionedThroughput": 1024},
+            "When specifying 'provisioned_throughput', the 'throughput_mode' must be set to 'provisioned'",
+        ),
+        (
+            {"ThroughputMode": "provisioned"},
+            "When specifying 'throughput_mode' to 'provisioned', the 'provisioned_throughput' option must be specified",
+        ),
+    ],
+)
+def test_efs_validator(section_dict, expected_error):
```
Comment (enrico-usai) [inline cli/tests/pcluster/schemas/test_schema_validators.py:31]: note: when moving a validator let's add a # FIXME moved comment in the original one, in this way it will be clear what we already moved in the new logic and what is missing.

File: cli/src/pcluster/config/cluster_config.py [hunk 18]
```diff
@@ -37,9 +39,10 @@ def __init__(self, validator_class: Validator, priority: int = 1, **kwargs):
 class Config(ABC):
     """Represent an abstract Configuration entity."""
 
-    def __init__(self):
+    def __init__(self, raise_on_error=False):
```
Comment (enrico-usai) [inline cli/src/pcluster/config/cluster_config.py:42]: I think we should remove the `raise_on_error` from the `Config` objects,
instead we can pass it to the `validate()` function and it will be propagated to all the validators.

It is an option related to the validators not to the config.

File: cli/tests/pcluster/validators/test_fsx.py [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+                "PerUnitStorageThroughput": 12,
+            },
+            "Capacity for FSx PERSISTENT HDD 12 MB/s/TiB file systems is increments of 6,000 GiB",
+        ),
+    ],
+)
+def test_fsx_storage_capacity_validator(section_dict, expected_error):
+    with pytest.raises(ConfigValidationError, match=expected_error):
+        fsx_config = FsxSchema().load(section_dict)
+        fsx_config.raise_on_error = True
```
Comment (enrico-usai) [inline cli/tests/pcluster/validators/test_fsx.py:63]: As said before instead of setting raise_on_error in teh config we can pass the parameter to the validate() function below.

File: cli/src/pcluster/schemas/cluster_schema.py [hunk 20]
```diff
@@ -129,135 +415,205 @@ def make_config(self, data, **kwargs):
...
+    @post_load
+    def make_config(self, data, **kwargs):
+        """Generate config object."""
+        return AdditionalIamPolicyConfig(**data)
+
+
+class IamSchema(BaseSchema):
+    """Represent the schema of IAM."""
+
+    roles = fields.Nested(RolesSchema, data_key="Roles")
```
Comment (enrico-usai) [inline cli/src/pcluster/schemas/cluster_schema.py:583]: minor: you can avoid the data_key

File: cli/src/pcluster/schemas/cluster_schema.py [hunk 21]
```diff
@@ -45,80 +109,302 @@ def evaluate_dynamic_defaults(self, raw_data, **kwargs):
...
+    @validates("mount_dir")
+    def validate_not_none_dir(self, value):
+        """Validate that user is not specifying /NONE or NONE as MountDir for any filesystem."""
+        if re.match("^/?NONE$", value):
+            raise ValidationError("{0} cannot be used as a mount directory".format(value))
+
+    @validates_schema
+    def validate_existence_of_settings(self, data, **kwargs):
+        """Validate that there is a setting."""
+        pass
```
Comment (enrico-usai) [inline cli/src/pcluster/schemas/cluster_schema.py:284]: let's add a # TODO

File: cli/src/pcluster/schemas/cluster_schema.py [hunk 22]
```diff
@@ -45,80 +109,302 @@ def evaluate_dynamic_defaults(self, raw_data, **kwargs):
...
+
+    @validates_schema
+    def validate_only_one_settings(self, data, **kwargs):
+        """Validate that there is no more than one settings."""
```
Comment (enrico-usai) [inline cli/src/pcluster/schemas/cluster_schema.py:288]: let's add a # TODO

File: cli/src/pcluster/schemas/cluster_schema.py [hunk 23]
```diff
@@ -45,80 +109,302 @@ def evaluate_dynamic_defaults(self, raw_data, **kwargs):
...
+    kms_key_id = fields.Str()
+    file_system_id = fields.Str(validate=validate.Regexp(r"^fs-[0-9a-z]{17}$"))
+    auto_import_policy = fields.Str(validate=validate.OneOf(["NEW", "NEW_CHANGED"]))
+    drive_cache_type = fields.Str(validate=validate.OneOf(["READ"]))
+    storage_type = fields.Str(validate=validate.OneOf(["HDD", "SSD"]))
+
+    @validates("imported_file_chunk_size")
+    def fsx_imported_file_chunk_size_validator(self, value):
+        """Validate ImportedFileChunkSize."""
+        if not 1 <= value <= 512000:
```
Comment (enrico-usai) [inline cli/src/pcluster/schemas/cluster_schema.py:253]: You can use: [validate.Range](https://marshmallow.readthedocs.io/en/stable/marshmallow.validate.html#marshmallow.validate.Range) by passing a custom error message.

File: cli/src/pcluster/schemas/cluster_schema.py [hunk 24]
```diff
@@ -45,80 +109,302 @@ def evaluate_dynamic_defaults(self, raw_data, **kwargs):
...
+        return EfsConfig(**data)
+
+    @validates_schema
+    def validate_existence_of_mode_throughput(self, data, **kwargs):
+        """Validate the conditional existence requirement between throughput_mode and provisioned_throughput."""
+        throughput_mode = data.get("throughput_mode")
+        provisioned_throughput = data.get("provisioned_throughput")
+        if throughput_mode != "provisioned" and provisioned_throughput:
+            raise ValidationError(
+                message="When specifying 'provisioned_throughput', the 'throughput_mode' must be set to 'provisioned'",
```
Comment (enrico-usai) [inline cli/src/pcluster/schemas/cluster_schema.py:218]: the message should be cleaned up to talk about provisioned througput and not "provisioned_througput" with the underscore in the middle that is the name of the old configuration parameter.

File: cli/src/pcluster/schemas/cluster_schema.py [hunk 25]
```diff
@@ -13,29 +13,93 @@
...
+def _camelcase(s):
+
+    parts = iter(s.split("_"))
+    return "".join(i.title() for i in parts)
 
 
 class BaseSchema(Schema):
     """Represent a base schema, containing all the features required by all the Schema classes."""
 
+    def on_bind_field(self, field_name, field_obj):
```
Comment (enrico-usai) [inline cli/src/pcluster/schemas/cluster_schema.py:92]: This is a great improvement.

File: cli/src/pcluster/schemas/cluster_schema.py [hunk 26]
```diff
@@ -45,80 +109,302 @@ def evaluate_dynamic_defaults(self, raw_data, **kwargs):
...
+    """Represent the generic SharedStorage schema."""
+
+    mount_dir = fields.Str(required=True, validate=_get_field_validator("file_path"))
+    ebs_settings = fields.Nested(EbsSchema, data_key="EBS")
+    efs_settings = fields.Nested(EfsSchema, data_key="EFS")
+    fsx_settings = fields.Nested(FsxSchema, data_key="FSX")
+
+    @post_load
+    def make_config(self, data, **kwargs):
+        """Return the right SharedStorage type object, according to the given type."""
```
Comment (enrico-usai) [inline cli/src/pcluster/schemas/cluster_schema.py:272]: The comment seems not aligned to the code.

File: cli/src/pcluster/schemas/cluster_schema.py [hunk 27]
```diff
@@ -45,80 +109,302 @@ def evaluate_dynamic_defaults(self, raw_data, **kwargs):
...
+        return SharedStorageConfig(**data)
+
+    @validates("mount_dir")
+    def validate_not_none_dir(self, value):
+        """Validate that user is not specifying /NONE or NONE as MountDir for any filesystem."""
+        if re.match("^/?NONE$", value):
```
Comment (enrico-usai) [inline cli/src/pcluster/schemas/cluster_schema.py:278]: you can use a [validate.Regex](https://marshmallow.readthedocs.io/en/stable/marshmallow.validate.html#marshmallow.validate.Regexp) by passing a custom error.

File: cli/src/pcluster/validators/fsx.py [hunk 28]
```diff
@@ -34,3 +69,131 @@ def __call__(self, import_path, imported_file_chunk_size, export_path, auto_impo
                 "When specifying auto import policy, the import path option must be specified",
                 FailureLevel.CRITICAL,
             )
+
+    def _validate_persistent_options(self, deployment_type, kms_key_id, per_unit_storage_throughput):
+        if deployment_type == "PERSISTENT_1":
+            if not per_unit_storage_throughput:
+                self._add_failure(
+                    "'per_unit_storage_throughput' must be specified when 'deployment_type = PERSISTENT_1'",
```
Comment (enrico-usai) [inline cli/src/pcluster/validators/fsx.py:77]: let's rememeber to cleanup all the validation messages. We can do it in a separate PR.

File: cli/src/pcluster/config/cluster_config.py [hunk 29]
```diff
@@ -65,49 +70,279 @@ def __repr__(self):
...
+        volume_type: str = None,
+        iops: int = None,
+        size: int = None,
+        encrypted: bool = None,
+        kms_key_id: str = None,
+        throughput: int = None,
+    ):
         super().__init__()
-        self.os = os
+        if volume_type is None:
```
Comment (enrico-usai) [inline cli/src/pcluster/config/cluster_config.py:87]: For the simplest conditions you can have a single row for the initialization:
```python
self.volume_type = MarkedStr("gp2") if volume_type is None else volume_type
```

In any case I'd keep the conditions and the initialization of the attribute grouped together.

File: cli/src/pcluster/config/cluster_config.py [hunk 30]
```diff
@@ -65,49 +70,279 @@ def __repr__(self):
...
+        self.daily_automatic_backup_start_time = daily_automatic_backup_start_time
+        self.per_unit_storage_throughput = per_unit_storage_throughput
+        self.backup_id = backup_id
+        self.kms_key_id = kms_key_id
+        self.file_system_id = file_system_id
+        self.auto_import_policy = auto_import_policy
+        self.drive_cache_type = drive_cache_type
+        self.storage_type = storage_type
+        self._register_validator(FsxValidator, fsx_config=self)
+        # TODO register validators
```
Comment (enrico-usai) [inline cli/src/pcluster/config/cluster_config.py:234]: you can change this todo to "split validators"

File: cli/src/pcluster/config/cluster_config.py [hunk 31]
```diff
@@ -65,49 +70,279 @@ def __repr__(self):
...
+# ---------------------- Networking ---------------------- #
+class ProxyConfig(Config):
+    """Represent the proxy."""
+
+    def __init__(self, http_proxy_address: str = None):
+        super().__init__()
+        self.http_proxy_address = http_proxy_address
+
+
+class BaseNetworkingConfig(Config):
```
Comment (enrico-usai) [inline cli/src/pcluster/config/cluster_config.py:266]: Base --> Common ?

File: cli/src/pcluster/config/cluster_config.py [hunk 32]
```diff
@@ -65,49 +70,279 @@ def __repr__(self):
...
+
+# ---------------------- Nodes ---------------------- #
+class ImageConfig(Config):
+    """Represent the configuration of an Image."""
+
+    def __init__(self, os: str, custom_ami: str = None):
+        super().__init__()
+        self.os = os
+        self.custom_ami = custom_ami
+        self._validators = []
```
Comment (enrico-usai) [inline cli/src/pcluster/config/cluster_config.py:345]: this initialization  is already done in the BaseConfig you can avoid it if there are no validators to register.

File: cli/src/pcluster/config/cluster_config.py [hunk 33]
```diff
@@ -116,23 +351,29 @@ class HeadNodeConfig(Config):
...
-        self.image_config = image_config
-        self.networking_config = networking_config
-        self.ssh_config = ssh_config
-        self._register_validator(InstanceTypeValidator, priority=100, instance_type=self.instance_type)
+        self.image = image
+        self.networking = networking
+        self.ssh = ssh
+        self.storage = storage
+        self.dcv = dcv
+        self._register_validator(InstanceTypeValidator, priority=1, instance_type=self.instance_type)
```
Comment (enrico-usai) [inline cli/src/pcluster/config/cluster_config.py:367]: not related to your patch but I'd rename _register_validator to _add_validator to be more compact. You can do it with IDE refactor utility.

File: cli/src/pcluster/config/cluster_config.py [hunk 34]
```diff
@@ -14,48 +14,90 @@
...
+    def __new__(cls, s: str, implied: bool = True):
+        """Create str object and add attributes."""
+        obj = str.__new__(cls, s)
+        obj.implied = implied
+        return obj
+
+
+class MyBool(object):
+    """
+    Create int object and add attributes.
```
[the remark this answers is not in the document]
Comment (hanwen-pcluste) [reply cli/src/pcluster/config/cluster_config.py]: Typo. Fixed

File: cli/src/pcluster/config/cluster_config.py [hunk 35]
```diff
@@ -14,54 +14,114 @@
...
+from pcluster.constants import EBS_VOLUME_TYPE_IOPS_DEFAULT
 from pcluster.validators.common import ValidationResult, Validator
+from pcluster.validators.ebs import EbsVolumeIopsValidator, EbsVolumeThroughputValidator, EbsVolumeTypeSizeValidator
 from pcluster.validators.ec2 import InstanceTypeValidator
 from pcluster.validators.fsx import FsxS3OptionsValidator
 
 
+# Extend build-in classes to record if a value is implied.
+# This is useful for distinguishing values implied vs specified by a user.
+class MyInt(int):
```
[the remark this answers is not in the document]
Comment (hanwen-pcluste) [reply cli/src/pcluster/config/cluster_config.py]: Done

File: cli/src/pcluster/config/cluster_config.py (hunk 9 above)
Comment (hanwen-pcluste) [reply cli/src/pcluster/config/cluster_config.py]: After discussion, we restored the name to priority instead of dependency.

File: cli/src/pcluster/config/cluster_config.py (hunk 18 above)
Comment (hanwen-pcluste) [reply cli/src/pcluster/config/cluster_config.py]: Done

File: cli/src/pcluster/config/cluster_config.py (hunk 29 above)
Comment (hanwen-pcluste) [reply cli/src/pcluster/config/cluster_config.py:86]: Yes. I took the current approach to group the defaults together. With the simplest conditions, things will mix together.
Here I am wasting some lines of code for better readability.

File: cli/src/pcluster/config/cluster_config.py (hunk 30 above)
Comment (hanwen-pcluste) [reply cli/src/pcluster/config/cluster_config.py]: Done

File: cli/src/pcluster/config/cluster_config.py (hunk 32 above)
Comment (hanwen-pcluste) [reply cli/src/pcluster/config/cluster_config.py]: Done

File: cli/src/pcluster/config/cluster_config.py (hunk 33 above)
Comment (hanwen-pcluste) [reply cli/src/pcluster/config/cluster_config.py]: Done

File: cli/src/pcluster/schemas/cluster_schema.py (hunk 25 above)
Comment (hanwen-pcluste) [reply cli/src/pcluster/schemas/cluster_schema.py:92]: Thank you. I was desperate before having this method. Haha

File: cli/src/pcluster/schemas/cluster_schema.py (hunk 24 above)
Comment (hanwen-pcluste) [reply cli/src/pcluster/schemas/cluster_schema.py]: I changed this one. There are some remaining messages to be cleaned up. I will change them once I have time.

File: cli/src/pcluster/schemas/cluster_schema.py (hunk 23 above)
Comment (hanwen-pcluste) [reply cli/src/pcluster/schemas/cluster_schema.py]: Good catch. Done

File: cli/src/pcluster/schemas/cluster_schema.py (hunk 26 above)
Comment (hanwen-pcluste) [reply cli/src/pcluster/schemas/cluster_schema.py]: Done

File: cli/src/pcluster/schemas/cluster_schema.py (hunk 27 above)
Comment (hanwen-pcluste) [reply cli/src/pcluster/schemas/cluster_schema.py:274]: I don't know a easy way to deny a pattern (instead of accepting a pattern) using validate.Regex

File: cli/src/pcluster/schemas/cluster_schema.py (hunk 21 above)
Comment (hanwen-pcluste) [reply cli/src/pcluster/schemas/cluster_schema.py]: Good catch. Instead of adding TODO, I implemented the method.

File: cli/src/pcluster/schemas/cluster_schema.py (hunk 22 above)
Comment (hanwen-pcluste) [reply cli/src/pcluster/schemas/cluster_schema.py]: The validation is implemented as said in the previous comment

File: cli/src/pcluster/schemas/cluster_schema.py (hunk 20 above)
Comment (hanwen-pcluste) [reply cli/src/pcluster/schemas/cluster_schema.py]: Done, also cleaned up similar issues elsewhere

File: cli/tests/pcluster/schemas/test_schema_validators.py (hunk 17 above)
Comment (hanwen-pcluste) [reply cli/tests/pcluster/schemas/test_schema_validators.py:31]: Ok

File: cli/tests/pcluster/validators/test_fsx.py (hunk 19 above)
Comment (hanwen-pcluste) [reply cli/tests/pcluster/validators/test_fsx.py]: Done
